bin/parallel.py

```diff
@@ -31,15 +33,27 @@ def __init__(self, f, num_threads=True):
         self.finish = False
 
         if self.num_threads:
+            if pin:
+                # only use available cores
+                cores = list(os.sched_getaffinity(0))
+                if self.num_threads > len(cores):
+                    self.num_threads = len(cores)
+
+                # sort cores by reversed id. If num_threads << len(cores) this ensures that we 
+                # use different physical cores instead of hyperthreads
+                num_bits = max(map(lambda c: c.bit_length(), cores))
+                cores.sort(key=lambda c: util.bit_reverse(c, num_bits))
```

Comment:
so on my 6-core 12-thread machine, i think 1 and 7 are on the same core, and 1..=6 are on distinct cores. 
I also always assumed hyperthreads would be consecutive but apparently not. (You can confirm this using `cat /proc/cpuinfo` or so i think.)

Reply:
hm you seem to be right...